desc/builder/file.go

```diff
@@ -575,9 +575,9 @@ func (fb *FileBuilder) TryAddService(sb *ServiceBuilder) error {
 // an ExtensionRegistry, when building the file.
 func (fb *FileBuilder) AddDependency(dep *FileBuilder) *FileBuilder {
 	if fb.explicitDeps == nil {
-		fb.explicitDeps = map[*FileBuilder]struct{}{}
+		fb.explicitDeps = map[string]*FileBuilder{}
 	}
-	fb.explicitDeps[dep] = struct{}{}
+	fb.explicitDeps[dep.GetName()] = dep
```

Comment:
This is not safe or correct. We are possibly overwriting a dependency that had different contents. Which means we can create a result that cannot be built/linked, but for different reasons.

For example, let's say there were two file descriptors, both for path `foo/bar.proto`. One of them includes a message type `Baz`. The other does not, but instead has a message type `Frobnitz`. If we try to refer to both message types, we cannot build/link the result because this will basically only link in the later one, so there is no symbol `Baz` available at link time.

Reply:
Added check.